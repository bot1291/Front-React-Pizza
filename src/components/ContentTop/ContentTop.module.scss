@import '../../styles/variables.scss';

.contentTop {
	display: flex;
	align-items: center;
	justify-content: space-between;
}