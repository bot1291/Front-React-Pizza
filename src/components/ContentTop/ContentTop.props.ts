import { DetailedHTMLProps, HTMLAttributes } from 'react';

export interface ContentTopProps
	extends DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement> {}
