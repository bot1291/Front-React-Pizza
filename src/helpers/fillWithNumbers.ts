const fillWithNumbers = (count: number): number[] => {
	const numbers: number[] = [];
	for (let i = 0; i < count; i++) {
		numbers.push(i);
	}
	return numbers;
};

export default fillWithNumbers;
