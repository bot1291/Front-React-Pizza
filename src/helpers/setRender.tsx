import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createReduxStore } from '../store/store';

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
const setRender = <T,>(Component: JSX.Element, store?: Record<string, T>) => {
	return render(
		<Provider store={createReduxStore(store)}>{Component}</Provider>
	);
};

export default setRender;
