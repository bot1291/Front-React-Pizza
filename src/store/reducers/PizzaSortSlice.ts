import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { TypeParams } from '../../interfaces/TypeParams';
import { IPizza } from '../../interfaces/IPizza';
import { fetchPizzas } from './data';

const allPizzaTypes = fetchPizzas.map((p) => p.types).flat();
const sortedTypes: string[] = [];
allPizzaTypes.forEach((p) => {
	if (!sortedTypes.includes(p)) {
		sortedTypes.push(p);
	}
});

interface IUserState {
	isLoading: boolean;
	error: string;
	pizzas: IPizza[];
	pizzasBackup: IPizza[];
	allPizzaTypes: string[];
	currentType: string;
	searchValue: string;
	currentSortParam: TypeParams;
	isSortedByType: boolean;
	isSortedToDown: boolean;
}

export const initialState: IUserState = {
	isLoading: false,
	error: '',
	pizzas: localStorage.pizzas
		? (JSON.parse(localStorage.pizzas) as unknown as IPizza[])
		: fetchPizzas,
	pizzasBackup: localStorage.pizzasBackup
		? JSON.parse(localStorage.pizzasBackup)
		: fetchPizzas,
	allPizzaTypes: sortedTypes,
	currentType: 'все',
	searchValue: '',
	isSortedByType: false,
	isSortedToDown: true,
	currentSortParam: {
		title: 'популярности',
		param: 'rating',
	},
};

export type PossibleSortParams = 'price' | 'title' | 'rating';

const checkForSimilarity = (word: string, search: string): boolean => {
	return word
		.toLowerCase()
		.split(' ')
		.join('')
		.includes(search.toLowerCase().split(' ').join(''));
};

export const pizzaSortSlice = createSlice({
	name: 'pizza',
	initialState,
	reducers: {
		setParam: (state, action: PayloadAction<IPizza[]>) => {
			state.pizzas = action.payload;
			state.pizzasBackup = state.pizzasBackup.map((pizzaBack) => {
				const replacementPizza = action.payload.find(
					(pizza) => pizzaBack.title === pizza.title
				);
				if (replacementPizza) {
					return replacementPizza;
				}
				return pizzaBack;
			});
		},
		sortToUpOrDown: (state, action: PayloadAction<boolean>) => {
			state.pizzas = state.pizzas.sort((a, b) => {
				if (
					a[state.currentSortParam.param] >
					b[state.currentSortParam.param]
				) {
					return action.payload ? -1 : 1;
				}
				return action.payload ? 1 : -1;
			});
			state.isSortedToDown = !state.isSortedToDown;
		},
		sortByParam: (state, action: PayloadAction<TypeParams>) => {
			state.pizzas = state.pizzas.sort((a, b) =>
				a[action.payload.param] > b[action.payload.param] ? 1 : -1
			);
			state.currentSortParam = action.payload;
			state.pizzasBackup = state.pizzasBackup.sort((a, b) =>
				a[action.payload.param] > b[action.payload.param] ? 1 : -1
			);
			state.isSortedToDown = true;
		},
		sortByType: (state, action: PayloadAction<string>) => {
			state.pizzas = state.pizzasBackup.filter((p) =>
				p.types.includes(action.payload)
			);
			state.currentType = action.payload;
			state.searchValue = '';
			state.isSortedByType = true;
		},
		unset: (state) => {
			state.pizzas = state.pizzasBackup;
			state.currentType = 'все';
			state.searchValue = '';
			state.isSortedByType = false;
		},
		search: (state, action: PayloadAction<string>) => {
			state.pizzas = (
				state.searchValue ? state.pizzasBackup : state.pizzas
			).filter(
				(p) =>
					checkForSimilarity(p.title, action.payload) ||
					checkForSimilarity(p.types.join(''), action.payload)
			);
		},
		setSearchValue: (state, action: PayloadAction<string>) => {
			state.searchValue = action.payload;
		},
	},
});

export default pizzaSortSlice.reducer;
