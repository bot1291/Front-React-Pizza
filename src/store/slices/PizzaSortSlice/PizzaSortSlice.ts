import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { TypeParams } from '../../../interfaces/TypeParams';
import { IPizza } from '../../../interfaces/IPizza';
import { fetchPizzas } from '../ActionCreators';
import checkForSimilarity from '../../../helpers/checkForSimilarity';

export interface IUserState {
	isLoading: boolean;
	error: boolean;
	pizzas: IPizza[];
	pizzasBackup: IPizza[];
	allPizzaTypes: string[];
	currentType: string;
	searchValue: string;
	currentSortParam: TypeParams;
	isSortedToDown: boolean;
}

export const initialState: IUserState = {
	isLoading: true,
	error: false,
	pizzas: [],
	pizzasBackup: [],
	allPizzaTypes: [],
	currentType: 'все',
	searchValue: '',
	isSortedToDown: localStorage.isDown === 'true' ? true : false,
	currentSortParam: {
		title: 'популярности',
		param: 'rating',
	},
};

export const pizzaSortSlice = createSlice({
	name: 'pizza',
	initialState,
	reducers: {
		setParam: (state, action: PayloadAction<IPizza[]>) => {
			state.pizzas = action.payload;
			state.pizzasBackup = state.pizzasBackup.map((pizzaBack) => {
				const replacementPizza = action.payload.find(
					(pizza) => pizzaBack.title === pizza.title
				);
				if (replacementPizza) {
					return replacementPizza;
				}
				return pizzaBack;
			});
		},
		sortToUpOrDown: (state, action: PayloadAction<boolean>) => {
			state.pizzas = state.pizzas.sort((a, b) => {
				if (
					a[state.currentSortParam.param] >
					b[state.currentSortParam.param]
				) {
					return action.payload ? -1 : 1;
				}
				return action.payload ? 1 : -1;
			});
			state.isSortedToDown = !action.payload;
		},
		sortByParam: (state, action: PayloadAction<TypeParams>) => {
			state.pizzas = state.pizzas.sort((a, b) =>
				a[action.payload.param] > b[action.payload.param] ? 1 : -1
			);
			state.pizzasBackup = state.pizzasBackup.sort((a, b) =>
				a[action.payload.param] > b[action.payload.param] ? 1 : -1
			);
			state.currentSortParam = action.payload;
		},
		sortByType: (state, action: PayloadAction<string>) => {
			state.pizzas = state.pizzasBackup.filter((p) =>
				p.types.includes(action.payload)
			);
			state.currentType = action.payload;
			state.searchValue = '';
		},
		unset: (state) => {
			state.pizzas = state.pizzasBackup;
			state.currentType = 'все';
			state.searchValue = '';
		},
		search: (state, action: PayloadAction<string>) => {
			state.pizzas = state.pizzasBackup.filter(
				(p) =>
					checkForSimilarity(p.title, action.payload) ||
					checkForSimilarity(p.types.join(''), action.payload)
			);
		},
		setSearchValue: (state, action: PayloadAction<string>) => {
			state.searchValue = action.payload;
		},
	},
	extraReducers: (builder) => {
		builder.addCase(fetchPizzas.fulfilled, (state, action) => {
			state.isLoading = false;
			state.error = false;
			state.pizzas = action.payload?.pizzas ?? [];
			state.allPizzaTypes = action.payload?.types ?? [];
			state.pizzasBackup = action.payload?.pizzas ?? [];
		});
		// eslint-disable-next-line @typescript-eslint/no-unused-vars
		builder.addCase(fetchPizzas.pending, (state, _action) => {
			state.isLoading = true;
		});
		// eslint-disable-next-line @typescript-eslint/no-unused-vars
		builder.addCase(fetchPizzas.rejected, (state, _action) => {
			state.isLoading = false;
			state.error = true;
		});
	},
});

export default pizzaSortSlice.reducer;

export const {
	search,
	setParam,
	setSearchValue,
	sortByParam,
	sortByType,
	sortToUpOrDown,
	unset,
} = pizzaSortSlice.actions;
