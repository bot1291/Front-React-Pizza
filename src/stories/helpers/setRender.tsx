import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { createReduxStore } from '../../store/store';

const setRender = <T,>(
	Component: JSX.Element,
	store?: Record<string, T>
): JSX.Element => {
	return (
		<MemoryRouter>
			<Provider store={createReduxStore(store)}>{Component}</Provider>
		</MemoryRouter>
	);
};

export default setRender;
