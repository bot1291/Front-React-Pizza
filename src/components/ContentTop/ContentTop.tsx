import styles from './ContentTop.module.scss';
import { ContentTopProps } from './ContentTop.props';
import cn from 'classnames';
import { FC } from 'react';
import { Sort } from '../Sort/Sort';
import { useAppSelector } from '../../hooks/redux';
import { CategoriesSkeleton, Categories } from '..';
import sortParams from '../../helpers/sortParams';

export const ContentTop: FC<ContentTopProps> = ({ className, ...props }) => {
	const { error, isLoading } = useAppSelector(
		(state) => state.pizzaSortReducer
	);

	return (
		<div className={cn(className, styles.contentTop)} {...props}>
			{isLoading || error ? (
				<CategoriesSkeleton data-testid="skeleton-wrapper" count={4} />
			) : (
				<>
					<Categories data-testid="categories-wrapper" />
					<Sort data-testid="sort-wrapper" sortParams={sortParams} />
				</>
			)}
		</div>
	);
};
