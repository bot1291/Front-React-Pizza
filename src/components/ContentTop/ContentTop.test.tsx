import setRender from '../../helpers/setRender';
import { ContentTop } from './ContentTop';
import { screen } from '@testing-library/dom';
import * as fillWithNumbers from '../../helpers/fillWithNumbers';
import { IUserState } from '../../store/slices/PizzaSortSlice/PizzaSortSlice';

const sortInitialState: IUserState = {
	isLoading: false,
	error: false,
	pizzas: [],
	pizzasBackup: [],
	allPizzaTypes: [],
	currentType: 'все',
	searchValue: '',
	isSortedToDown: true,
	currentSortParam: {
		title: 'цене',
		param: 'currentPrice',
	},
};

describe('ContentTop', () => {
	it('Should display data after loading', async () => {
		setRender(<ContentTop />, {
			pizzaSortReducer: sortInitialState,
		});
		const categoriesWrapper = await screen.findByTestId(
			'categories-wrapper'
		);
		const sortWrapper = await screen.findByTestId('sort-wrapper');
		expect(categoriesWrapper).toBeInTheDocument();
		expect(sortWrapper).toBeInTheDocument();
	});

	it('If happened error display skeleton', () => {
		const spyFunc = jest.spyOn(fillWithNumbers, 'default');
		setRender(<ContentTop />, {
			pizzaSortReducer: { ...sortInitialState, isLoading: true },
		});
		const categorieSkeletonWrapper = screen.getByTestId('skeleton-wrapper');
		const categoryElem = screen.getAllByTestId('category');
		expect(spyFunc).toHaveBeenCalled();
		expect(categorieSkeletonWrapper).toBeInTheDocument();
		expect(categoryElem).toHaveLength(4);
		expect(categorieSkeletonWrapper).toMatchSnapshot();
	});
});
