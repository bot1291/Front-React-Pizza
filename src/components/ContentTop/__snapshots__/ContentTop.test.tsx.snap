// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays categories and sort table if loaded If data is loading or threw error display skeleton 1`] = `
<div
  class="categories"
  data-testid="skeleton-wrapper"
>
  <div
    class="category"
    data-testid="category"
  >
    <div
      class="title"
    />
  </div>
  <div
    class="category"
    data-testid="category"
  >
    <div
      class="title"
    />
  </div>
  <div
    class="category"
    data-testid="category"
  >
    <div
      class="title"
    />
  </div>
  <div
    class="category"
    data-testid="category"
  >
    <div
      class="title"
    />
  </div>
  <div
    class="active category"
  >
    <div
      class="title"
    />
  </div>
</div>
`;
